src/mk/compact.rs

```diff
@@ -0,0 +1,290 @@
+use crate::utils::{crypto::*, num::is_even};
+
+pub struct Node<T> {
+    value: T,
+}
+
+impl<T> Clone for Node<T>
+where
+    T: Clone,
+{
+    fn clone(&self) -> Self {
+        Node {
+            value: self.value.clone(),
+        }
+    }
+}
+
+type MKNode = Node<Hash>;
+
+pub struct CompactMerkleTree {
+    leaves: Vec<MKNode>,
+    root_hash: Hash,
+}
+
+impl CompactMerkleTree {
+    fn create<T: DataToHash>(data: &[T]) -> Option<Self> {
+        if data.is_empty() {
+            return None;
+        }
+        let leaves: Vec<Node<[u8; 64]>> = Self::get_leaves_from(data);
+        let root_hash = Self::calculate_root(leaves.clone());
+        Some(Self { leaves, root_hash })
+    }
+
+    fn calculate_root(mut leaves: Vec<MKNode>) -> Hash {
+        while leaves.len() > 1 {
+            leaves = Self::get_parent_nodes(&leaves);
```

Comment:
Could leaves be borrowed by calculate_root? so you don't have to clone the leaves each time. I'm also not sure you need leaves to be mutable.
```rust
fn calculate_root(leaves: &Vec<MKNode>) -> Hash
```

Reply:
no you can't 😢, because you need to mutate it.